File: python_code/variable_selection/c2_Bayesopt_ann_for_varselect.py
import numpy as np
from skopt import gp_minimize
from skopt.space import Integer, Categorical
from sklearn.neural_network import MLPRegressor
from sklearn.model_selection import KFold, cross_val_score
import warnings
warnings.filterwarnings('ignore')


def SumSqr(params, XX, YY, cvss):
    n_layers = int(params[0])
    layer1 = int(params[1])
    layer2 = int(params[2])
    activation = params[3]

    if n_layers == 1:
        hidden_layer_sizes = (layer1,)
    else:
        hidden_layer_sizes = (layer1, layer2)

    act_map = {'tanh': 'tanh', 'sigmoid': 'logistic', 'relu': 'relu'}

    Mdl = MLPRegressor(
        hidden_layer_sizes=hidden_layer_sizes,
        activation=act_map.get(activation, 'relu'),
        max_iter=2000,
        random_state=1,
        early_stopping=True
    )

    Mdl.fit(XX, YY)

    SST = np.sum((YY - np.mean(YY))**2)
    y_pred = Mdl.predict(XX)
    SSEmdl = np.sum((YY - y_pred)**2)

    cv_scores = cross_val_score(Mdl, XX, YY, cv=cvss, scoring='neg_mean_squared_error')
    SSEcv = -np.mean(cv_scores) * len(YY) * 5
    R2CV = 1 - (SSEcv / SST)
    R2 = 1 - (SSEmdl / SST)

    output = {'R2': R2, 'R2CV': R2CV, 'Mdl': Mdl}
    target = 1 - R2CV
    return target, output


def Optopt():
    dimensions = [
        Integer(1, 2, name='NumLayers'),
        Integer(2, 10, name='Layer_1'),
        Integer(2, 10, name='Layer_2'),
        Categorical(['tanh', 'sigmoid', 'relu'], name='Activation')
    ]
    return dimensions


def c2_Bayesopt_ann_for_varselect(X, Y):
    numFolds = 5
    np.random.seed(1)

    kf = KFold(n_splits=numFolds, shuffle=True, random_state=1)
    cvss = list(kf.split(X))

    dimensions = Optopt()

    def objective(params):
        target, _ = SumSqr(params, X, Y, cvss)
        return target

    res = gp_minimize(
        objective,
        dimensions,
        acq_func='EI',
        n_calls=60,
        random_state=1,
        verbose=False
    )

    best_params = res.x
    target, output = SumSqr(best_params, X, Y, cvss)

    return output['R2'], output['R2CV'], output['Mdl']

File: variable_selection/c2_Bayesopt_ann_for_varselect.m
function output=c2_Bayesopt_ann_for_varselect (X,Y) 

    numFolds = 5;
    rng (1)
    cvs = cvpartition(height(X),'KFold',numFolds);
    results = bayesopt(@(paraz)SumSqr(paraz,X,Y,cvs),Optopt(),'ConditionalVariableFcn',@condition,'AcquisitionFunctionName','expected-improvement-plus',...
        'IsObjectiveDeterministic',true,'UseParallel',false,'MaxObjectiveEvaluations',60,'Verbose',1,'PlotFcn',[]);
    %,'ConditionalVariableFcn',@condition'AcquisitionFunctionName','expected-improvement-plus'
    A1=results.XAtMinObjective;
    [~,output]=SumSqr(A1,X,Y,cvs);
disp('')



function X = condition(X)
X.Layer_2(X.NumLayers ~= 2) = NaN;
disp('');


function [target,userdata]=SumSqr(z,XX,YY,cvss)
msz=[z.Layer_1 z.Layer_2];
msz=msz(~isnan(msz));

rng (1)
Mdl=fitrnet(XX,YY,'InitialStepSize','auto','Activations',char(z.Activation),'Standardize',1,'LayerSizes',msz);%'Lambda',z.(5),
Mdlcv=crossval(Mdl,'CVPartition',cvss);
% yHat = predict(Mdl,XX);
% MSE_Mdl=resubLoss(Mdl);
% mdl_R2=corr(predict(Mdl,Mdl.X),Mdl.Y)^2;
% CV_R2=corr(kfoldPredict(Mdlcv),Mdlcv.Y)^2;%%'CVPartition',cvss,

SST=sum((Mdl.Y-mean(Mdl.Y)).^2);
SSEmdl=resubLoss(Mdl)*numel(Mdl.Y);
SSEcv=kfoldLoss(Mdlcv)*numel(Mdlcv.Y);
R2CV=1-(SSEcv/SST);
R2 = 1-(SSEmdl/SST);


userdata={R2,R2CV,Mdl};
target=1-R2CV;





function vars=Optopt()
v1= optimizableVariable('NumLayers',[1 2],'Type','integer','Transform','none');
v2= optimizableVariable('Layer_1',[2 10],'Type','integer','Transform','log');
v3= optimizableVariable('Layer_2',[2 10],'Type','integer','Transform','log');
v4= optimizableVariable('Activation',{'tanh' 'sigmoid' 'relu'},'Type','categorical');

%v5= optimizableVariable('Regularization',[8.1301e-08 813.01],'Type','real','Transform','log');
vars=[v1 v2 v3 v4];

